### PR TITLE
[OSD-14200] added accountpool_validation controller

### PR DESCRIPTION
**Pull request by reedcort**
# What is being added?
_Is this a fix for a bug?  What's the bug?  Is this a new feature? Please describe it. Is this just a small typo fix?  That's fine too!_

## Checklist before requesting review

- [ ] I have tested this locally
- [x] I have included unit tests
- [ ] I have updated any corresponding documentation

## Steps To Manually Test
_Please provide us steps to reproduce the scenarios needed to test this.  If an integration test is provided, let us know how to run it. If not, enumerate the steps to validate this does what it's supposed to._
1. Start the operator
1. Run the thing
1. Observe X in the spec for the thing
1. Clean up the thing

Ref OSD-0000

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -277,6 +277,10 @@ predeploy-aws-account-operator: ## Predeploy AWS Account Operator
 	@ls deploy/crds/*.yaml | xargs -L1 oc apply -f
 	# Create zero size account pool
 	@oc apply -f hack/files/aws.managed.openshift.io_v1alpha1_zero_size_accountpool.yaml
+	#@oc process --local -p NAME="zero-size-accountpool" -p SIZE=1 -p TYPE="Default" -p QUOTA=10 -f hack/templates/aws.managed.openshift.io_v1alpha1_accountpool.tmpl | oc apply -f -
```
Comment (AlexVulaj) [inline Makefile:280]: Do we want to delete this?

File: controllers/validation/accountpool_validation_controller.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+}
+
+func (r *AccountPoolValidationReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	r.awsClientBuilder = &awsclient.Builder{}
+	maxReconciles, err := utils.GetControllerMaxReconciles(controllerName)
+	if err != nil {
+		log.Error(err, "missing max reconciles for controller", "controller", controllerName)
+	}
+
+	rwm := utils.NewReconcilerWithMetrics(r, accountpoolcontrollerName)
```
Comment (dkeohane) [inline controllers/validation/accountpool_validation_controller.go:183]: I think you can replace the string object accountpoolcontrollerName with just the string `accountpoolvalidation`

File: controllers/validation/accountpool_validation_controller.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var accountpoollog = logf.Log.WithName("controller_accountpoolvalidation")
+
+const (
+	accountpoolcontrollerName  = "accountpoolvalidation"
+	AccountPendingVerification = "PendingVerification"
```
Comment (dkeohane) [inline controllers/validation/accountpool_validation_controller.go:32]: We have this defined in another package, just pull it in from there

File: controllers/validation/accountpool_validation_controller.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+)
+
+type AccountPoolValidationReconciler struct {
+	Client           client.Client
+	Scheme           *runtime.Scheme
+	awsClientBuilder awsclient.IBuilder
+}
+
+func (r *AccountPoolValidationReconciler) accountSpecUpdate(reqLogger logr.Logger, account *awsv1alpha1.Account) error {
```
Comment (dkeohane) [inline controllers/validation/accountpool_validation_controller.go:41]: We have this defined in another package, just pull it in from there. If it's private, let's look to make it public

File: controllers/validation/accountpool_validation_controller.go [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+			account.Spec.RegionalServiceQuotas = parsedRegionalServiceQuotas
+			account.Status.RegionalServiceQuotas = make(awsv1alpha1.RegionalServiceQuotas)
+		}
+
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *AccountPoolValidationReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
```
Comment (dkeohane) [inline controllers/validation/accountpool_validation_controller.go:149]: Can we move Reconcile up to the top of the file

File: controllers/validation/accountpool_validation_controller.go [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+	data := make(map[string]AccountPool)
+	err = yaml.Unmarshal([]byte(accountpoolString), &data)
+
+	if err != nil {
+		reqLogger.Error(err, "Failed to unmarshal yaml")
+		return nil, err
+	}
+
+	var parsedRegionalServiceQuotas = make(awsv1alpha1.RegionalServiceQuotas)
+	// If the pool we've specified in the account doesn't exist, we need to error out
```
Comment (dkeohane) [inline controllers/validation/accountpool_validation_controller.go:107]: comment no longer relevant

File: controllers/validation/accountpool_validation_controller.go [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+// Validate Account ServiceQuota Matches AccountPool ServiceQuota
+// first get configmap ref accounpool_controller (handleservicequotas) and check if accountpool has servicequota defined in configmap
+// if not just return
+// if defined get all accounts for that accountpool
+// then check if theres a difference between servicequota in configmap and wants in account CR spec
+// if theres a difference take the configmap servicequota and drop in the account CR spec
+// wipe current account cr servicequota status
+// then call setCurrentAccountServiceQuotas
+func (r *AccountPoolValidationReconciler) checkAccountPoolServiceQuota(reqLogger logr.Logger, accountPoolName string) (awsv1alpha1.RegionalServiceQuotas, error) {
```
[the remark this answers is not in the document]
Comment (dkeohane) [reply controllers/validation/accountpool_validation_controller.go:77]: Also, a lot of this code was taken from elsewhere, can we find a way not to duplicate the same code and refactor?
DRY https://en.wikipedia.org/wiki/Don%27t_repeat_yourself :smiley_cat:

File: controllers/validation/accountpool_validation_controller.go [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+
+func (r *AccountPoolValidationReconciler) accountSpecUpdate(reqLogger logr.Logger, account *awsv1alpha1.Account) error {
+	err := r.Client.Update(context.TODO(), account)
+	if err != nil {
+		reqLogger.Error(err, fmt.Sprintf("Account spec update for %s failed", account.Name))
+	}
+	return err
+}
+
+func (r *AccountPoolValidationReconciler) getAccountPoolAccounts(name string) ([]awsv1alpha1.Account, error) {
```
Comment (dkeohane) [inline controllers/validation/accountpool_validation_controller.go:49]: Do we have this function defined elsewhere? If so, we can probably reuse

**Review by dkeohane — COMMENTED** [review]
Solid first start, just need to remove areas of potential duplication and some minor cleanup, then you can move on to handling the newly wiped account on the account validation controller side.

To note : I didn't look through any of your tests at this point

File: controllers/validation/account_validation_controller.go [hunk 9]
```diff
@@ -439,10 +452,91 @@ func (r *AccountValidationReconciler) Reconcile(ctx context.Context, request ctr
...
+		return utils.RequeueWithError(err)
+	}
+
+	// check if account belongs to accountpool
+	if !account.IsBYOC() {
+		if account.Spec.RegionalServiceQuotas != nil {
+			if account.Status.RegionalServiceQuotas == nil {
+				acct.SetCurrentAccountServiceQuotas(reqLogger, &account, regionsEnabledInAccount)
+			} else {
+				_, err := r.getServiceQuotaRequest(reqLogger, &account, awsSetupClient)
```
Comment (dkeohane) [inline controllers/validation/account_validation_controller.go:491]: You're not doing anything with the return values in this function, even though they're reconcile.Result - you can probably just return directly

File: controllers/account/account_controller_test.go [hunk 10]
```diff
@@ -1549,7 +1549,10 @@ var _ = Describe("Account Controller", func() {
 							},
 						},
 					}, nil)
-					err := setCurrentAccountServiceQuotas(r, nullLogger, account, mockAWSClient)
+					regionsEnabledInAccount, _ := mockAWSClient.DescribeRegions(&ec2.DescribeRegionsInput{
+						AllRegions: aws.Bool(false),
+					})
+					err := SetCurrentAccountServiceQuotas(nullLogger, account, regionsEnabledInAccount)
```
Comment (reedcort) [inline controllers/account/account_controller_test.go:1555]: Wasn't sure want to change this to. The original line :  `err := setCurrentAccountServiceQuotas(r, nullLogger, account, mockAWSClient)` .  The function was updated

Comment (bergmannf) [reply controllers/account/account_controller_test.go:1555]: The mock that is being used there is just returning the following `DescribeRegionsOutput`:

```
&ec2.DescribeRegionsOutput{
						Regions: []*ec2.Region{
							{
								RegionName: aws.String("us-east-1"),
							},
						},
					}, nil)
```

Given that this is now passed directly, you should be able to remove the mockClient completely from this test and just pass in the DescribeRegionsOutput directly.

File: controllers/validation/account_validation_controller.go [hunk 11]
```diff
@@ -439,10 +452,108 @@ func (r *AccountValidationReconciler) Reconcile(ctx context.Context, request ctr
...
+				}
+
+			}
+		}
+	}
 	return utils.DoNotRequeue()
 }
 
-// SetupWithManager sets up the controller with the Manager.
+func (r *AccountValidationReconciler) getServiceQuotaRequest(reqLogger logr.Logger, currentAcctInstance *awsv1alpha1.Account, awsSetupClient awsclient.Client) (reconcile.Result, error) {
```
Comment (dkeohane) [inline controllers/validation/account_validation_controller.go:511]: Can this entire function be used in HandleNonCCSPendingVerification as well by removing the dependency on AccountValidationReconciler which I think is only being used to call statusUpdate?

Comment (reedcort) [reply controllers/validation/account_validation_controller.go:511]: This function is being used in HandleNonCCSPendingVerification already.

File: controllers/validation/accountpool_validation_controller.go [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var logs = logf.Log.WithName("controller_accountpoolvalidation")
+
+const (
+	ControllerName = "accountpoolvalidation"
+)
```
Comment (reedcort) [inline controllers/validation/accountpool_validation_controller.go:29]: Had to change the name of these two variables in order to redeclare them for the accountpool_validation_controller.

**Comment by reedcort** [comment]
/hold

Before this PR can be merged updates need to be made to the "IsOwnedByAccountPool" function in account_types

**Pull request closed without merging**